src/gpu/optixPrograms.cu

```diff
@@ -266,17 +224,96 @@ extern "C" __global__ void __closesthit__()
 		Vec3f absPointVelocityInSensorFrame = ctx.rayOriginToWorld.rotation().inverse() * absPointVelocity;
 		Vec3f relPointVelocityBasedOnSensorLinearVelocity = absPointVelocityInSensorFrame - ctx.sensorLinearVelocityXYZ;
 
-		Vec3f hitRays = ctx.rayOriginToWorld.inverse() * hitWorld;
+		Vec3f hitRays = ctx.rayOriginToWorld.inverse() * hitWorldRaytraced;
 		Vec3f relPointVelocityBasedOnSensorAngularVelocity = Vec3f(.0f) - ctx.sensorAngularVelocityRPY.cross(hitRays);
 		relPointVelocity = relPointVelocityBasedOnSensorLinearVelocity + relPointVelocityBasedOnSensorAngularVelocity;
 
 		radialSpeed = hitRays.normalized().dot(relPointVelocity);
 	}
+	saveRayResult<true>(hitWorldSeenBySensor, distance, intensity, entityId, absPointVelocity, relPointVelocity, radialSpeed,
+	                    wNormal, incidentAngle);
+}
+
+extern "C" __global__ void __anyhit__() {}
+
+// Helper functions implementations
 
-	saveRayResult<true>(hitWorld, distance, intensity, objectID, absPointVelocity, relPointVelocity, radialSpeed, wNormal,
-	                    incidentAngle);
+__device__ void shootSamplingRay(const Mat3x4f& ray, float maxRange, unsigned int beamSampleRayIdx)
+{
+	Vec3f origin = ray * Vec3f{0, 0, 0};
+	Vec3f dir = ray * Vec3f{0, 0, 1} - origin;
+	const unsigned int flags = OPTIX_RAY_FLAG_DISABLE_ANYHIT; // TODO: try adding OPTIX_RAY_FLAG_CULL_BACK_FACING_TRIANGLES
+	optixTrace(ctx.scene, origin, dir, 0.0f, maxRange, 0.0f, OptixVisibilityMask(255), flags, 0, 1, 0, beamSampleRayIdx);
 }
 
-extern "C" __global__ void __miss__() { saveNonHitRayResult(ctx.farNonHitDistance); }
+__device__ Mat3x4f makeBeamSampleRayTransform(float halfDivergenceAngleRad, unsigned int circleIdx, unsigned int vertexIdx)
+{
+	if (ctx.beamHalfDivergence == 0.0f) {
+		return Mat3x4f::identity();
+	}
+	const auto circleDivergence = halfDivergenceAngleRad * (1.0f - static_cast<float>(circleIdx) / MULTI_RETURN_BEAM_CIRCLES);
+	auto vertexAngleStep = 2.0f * M_PIf / MULTI_RETURN_BEAM_VERTICES;
+	// First, rotate around X to move the ray vector (initially {0,0,1}) to diverge from the Z axis by circleDivergence.
+	// Then rotate around Z to move the ray vector on the circle.
+	return Mat3x4f::rotationRad(0, 0, static_cast<float>(vertexIdx) * vertexAngleStep) * // Second
+	       Mat3x4f::rotationRad(circleDivergence, 0, 0);
+}
 
-extern "C" __global__ void __anyhit__() {}
+__device__ void saveNonHitRayResult(float nonHitDistance)
+{
+	Mat3x4f ray = ctx.rays[optixGetLaunchIndex().x];
+	Vec3f origin = ray * Vec3f{0, 0, 0};
+	Vec3f dir = ray * Vec3f{0, 0, 1} - origin;
```

Comment:
Wouldn't be faster by using `optixGetWorldRayDirection()` and `optixGetWorldRayOrigin()`?

Reply:
It would be faster, but it does not work since this function is called from `__raygen__`, where those functions are illegal to call :oncoming_police_car: